components/board/client.go

```diff
@@ -245,14 +266,139 @@ func (dic *digitalInterruptClient) Tick(ctx context.Context, high bool, nanoseco
+func (dic *digitalInterruptClient) RemoveCallback(ch chan Tick) {
 	panic(errUnimplemented)
 }
 
+func (c *client) StreamTicks(ctx context.Context, interrupts []string, ch chan Tick, extra map[string]interface{}) error {
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	stream := &interruptStream{
+		extra:  ext,
+		client: c,
+	}
+
+	err = stream.startStream(ctx, interrupts, ch)
+
+	if err != nil {
+		return err
+	}
+
+	c.mu.Lock()
+	c.interruptStreams = append(c.interruptStreams, stream)
+	c.mu.Unlock()
+
+	return nil
+
+}
+
+func (s *interruptStream) startStream(ctx context.Context, interrupts []string, ch chan Tick) error {
+	s.streamMu.Lock()
+	defer s.streamMu.Unlock()
+
+	s.streamRunning = true
+	s.streamReady = make(chan bool, 1024)
+	s.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	s.cancelBackgroundWorkers = cancel
+
+	// Create a background go routine that connects to the server's stream
+	utils.PanicCapturingGo(func() {
+		defer s.activeBackgroundWorkers.Done()
+		s.connectTickStream(ctx, interrupts, ch)
+	})
+
+	//wait until the stream is ready to return
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.streamReady:
+		return nil
+
+	}
+}
+
+func (s *interruptStream) connectTickStream(ctx context.Context, interrupts []string, ch chan Tick) {
+	defer func() {
+		s.streamMu.Lock()
+		defer s.streamMu.Unlock()
+		s.streamCancel = nil
+		s.streamRunning = false
+	}()
+
+	streamCtx, cancel := context.WithCancel(ctx)
+	s.streamCancel = cancel
+
+	select {
+	case <-ctx.Done():
+		return
+	default:
+	}
+
+	req := &pb.StreamTicksRequest{
+		Name:     s.client.info.name,
+		PinNames: interrupts,
+	}
+
+	//call the server to start streaming ticks
+	stream, err := s.client.client.StreamTicks(streamCtx, req)
+	if err != nil {
+		s.client.logger.CError(ctx, err)
+		return
+	}
+
+	if s.streamReady != nil {
+		close(s.streamReady)
+	}
+	s.streamReady = nil
+
+	// repeatly recieve from the stream
+	for {
+		select {
+		case <-ctx.Done():
+			s.client.logger.CError(ctx, err)
+			s.closeStream()
+			return
+		default:
+		}
+		streamResp, err := stream.Recv()
+		if err != nil {
+			// only debug log the context canceled error
+			s.client.logger.Debug(err)
+			s.closeStream()
+			return
+		} else {
+			// If there is a response with a tick, send it to the channel
+			tick := Tick{
+				Name:             streamResp.PinName,
+				High:             streamResp.High,
+				TimestampNanosec: streamResp.Time,
+			}
+			ch <- tick
+		}
+	}
+}
+
+func (s *interruptStream) closeStream() {
+	s.streamCancel()
+
+	for i := range s.interruptStreams {
+		if s.client.interruptStreams[i] == s {
+			// To remove this item, we replace it with the last item in the list, then truncate the
+			// list by 1.
+			s.client.interruptStreams[i] = s.client.interruptStreams[len(s.client.interruptStreams)-1]
+			s.client.interruptStreams = s.client.interruptStreams[:len(s.client.interruptStreams)-1]
```

Comment:
This is not thread safe: if two streams on the same client are trying to each close themselves at the same time, we might end up with copies of the last stream in the list. For example:
1. The client has the list of streams [A, B, C, D]
2. Stream A begins to close itself, and mutates the client's list to [D, B, C, D]
3. Stream B begins to close itself, and mutates the client's list to [D, D, C, D]
4. Stream A finishes closing itself, and mutates the client's list to [D, D, C]
5. Stream B finishes closing itself, and mutates the client's list to [D, D]

Instead, I think all manipulation of stuff within `s.client` belongs in the client struct, locked behind the client's mutex. That approach would also fit better with the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter), which kinda says you should interact with the public sections of the things directly connected to you, but not the private implementations within any of them.